portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -22,44 +33,65 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some placeholder comment saying 'Hello, Omar!'. TODO(oumontiel): modify this file to handle comments data */
+/** Servlet that posts and retrieves comments from Datastore. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   /**
-   * List of messages to be fetched from this Data Servlet
-   * @private {ArrayList<String>}
+   * Gets the comments, in JSON format, taken from user input in the form from the HTML,
+   * which are stored in the messages variable.
    */
-  private ArrayList<String> messages;
-
   @Override
-  public void init() {
-    messages = new ArrayList<>();
-    messages.add("Once upon a time...");
-    messages.add("Twice upon two times...");
-    messages.add("Thrice upon three times...");
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    /** Create a query and prepare it with the data stored in Datastore. */
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery preparedQuery = datastore.prepare(query);
+
+    /** Set maximum number of comments to be included in the response. */
+    int numOfComments = Integer.parseInt(request.getParameter("limit"));
```

Comment:
What if this isn't an valid integer? I say this below, but each individual part of your feature should be robust to 1) changes or 2) an alternative implementation, in case someone were to reuse just one piece of your feature to build their own product.

Reply:
Should I initialize to 0, then validate 'Integer.parseInt(request.getParameter("limit"))' to be >=0, and if so update the value? (Same approach to previous comment)

Reply:
Look at what they do here - https://github.com/OUMontiel/my-portfolio/blob/master/walkthroughs/week-3-server/examples/subtraction-game/src/main/java/com/google/sps/servlets/SubtractionServlet.java - you should catch the error from parseInt and fallback to something, whether that's giving the user feedback that there's an error (preferable longterm, though I don't expect you to do that in this project given the time constraint) or falling back to a sensible value (which could be 0 or something else!)

Reply:
Thanks for the example! Really helped. Implemented it the same way, plus added a throw for when the value is less than 0. Hope I did this (and the last one) correctly.